posts/2019-07-04-Rust-1.36.0.md

```diff
@@ -0,0 +1,164 @@
+
+### NLL on Rust 2015
+
+[nll-2018]: https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html#non-lexical-lifetimes
+[soundness]: https://en.wikipedia.org/wiki/Soundness
+[felix-blog]: http://blog.pnkfx.org/blog/2019/06/26/breaking-news-non-lexical-lifetimes-arrives-for-everyone/
+[crater-nll]: https://github.com/rust-lang/rust/issues/60680#issuecomment-495089654
+
+[Previously in Rust 1.31.0][nll-2018] we told you about NLL (Non-Lexical Lifetimes), an improvement to the language that makes the borrow checker smarter and more user friendly. For example, you may now write:
+
+```rust
+fn main() {
+    let mut x = 5;
+    let y = &x;
+    let z = &mut x; // This was not allowed before 1.31.0.
+}
+```
+
+In 1.31.0 NLL was stabilized for Rust 2018 only with a promise that we would back port it to Rust 2015 as well. With Rust 1.36.0, we are happy to announce that we have done so! NLL is now available in Rust 2015 as well.
+
+With NLL on both editions, we are closer to removing the old borrow checker. However, the old borrow checker unfortunately accepted some [unsound][soundness] code it should not have. As a result, NLL is currently in a "migration mode" wherein we will emit warnings instead of errors if the NLL borrow checker rejects code the old AST borrow checker would accept. Please see [this list][crater-nll] to see if one of your crates are affected.
+
+To read more about NLL, MIR, the story around fixing soundness holes, and what you can do about the warnings if you have them, [read Felix Klock's blog post][felix-blog].
+
+### A new `HashMap<K, V>` implementation
+
+[`hashbrown`]: https://crates.io/crates/hashbrown
+[`HashMap<K, V>`]: https://doc.rust-lang.org/std/collections/struct.HashMap.html
+[pr-hashbrown]: https://github.com/rust-lang/rust/pull/58623
+[SwissTable]: https://abseil.io/blog/20180927-swisstables
+[pr-hashbrown-perf]: https://perf.rust-lang.org/compare.html?start=b57fe74a27590289fd657614b8ad1f3eac8a7ad2&end=abade53a649583e40ed07c26ee10652703f09b58&stat=wall-time
+
+In Rust 1.36.0, the `HashMap<K, V>` implementation has been [replaced][pr-hashbrown] with the one in the [`hashbrown`] crate which is based on the [SwissTable] design. While the interface is the same, the `HashMap<K, V>` implementation is now on [average faster][pr-hashbrown-perf] and has lower memory overhead. Note that unlike the `hashbrown` crate, the implementation in `std` continues to use SipHash 1-3.
+
+### [`--offline`] support in Cargo
+
+[`--offline`]: https://doc.rust-lang.org/cargo/commands/cargo-build.html#cargo_build_manifest_options
+[`cargo fetch`]: https://doc.rust-lang.org/cargo/commands/cargo-fetch.html
+
+During most builds, Cargo doesn't interact with the network. Sometimes, however, Cargo has to. Such is the case when a dependency is added and the latest compatible version needs to be downloaded. At times, network access is not an option though, for example on an airplane or in isolated build environments.
```

Comment:
Prior to this change, do _all_ new dependencies cause network access, at least in order to determine whether the locally cached version (if any) is up to date? If so, I would leave out the "latest compatible version" clause.

```suggestion
During most builds, Cargo doesn't interact with the network. Sometimes, however, it does need network access, for instance when a new dependency is added. But in some environments, such as an airplane or an isolated build environment, network access is not an option.
```

Reply:
I have no idea; r? @nrc

Reply:
No, Cargo would check the cached index. Nobody seems quite sure when that cached index must be updated, but that is what mostly causes network access without `--offline`. At the least, Cargo does it when it can't resolve a version.

I have a draft blog post about --offline which I'll publish at about the same time as the release (depending on how long it takes to finish the post :-) ).

Reply:
> Nobody seems quite sure when that cached index must be updated...

🤣